apps/interpreter/src/interpreter.ts

```diff
@@ -105,66 +105,112 @@ async function runPipeline(
 
   logPipelineOverview(pipeline, runtimeParameters, executionContext.logger);
 
-  const executionOrder: Array<{
-    block: BlockDefinition;
-    value: IOTypeImplementation | null;
-  }> = getBlocksInTopologicalSorting(pipeline).map((block) => {
-    return { block: block, value: NONE };
-  });
+  const startTime = new Date();
+
+  const executionOrder = getBlocksInTopologicalSorting(pipeline).map(
+    (block) => {
+      return { block: block, value: NONE };
+    },
+  );
+  const exitCode = await executeBlocks(executionContext, executionOrder);
+
+  logExecutionDuration(startTime, executionContext.logger);
+  return exitCode;
+}
+
+interface ExecutionOrderItem {
+  block: BlockDefinition;
+  value: IOTypeImplementation | null;
+}
+
+async function executeBlocks(
+  executionContext: ExecutionContext,
+  executionOrder: ExecutionOrderItem[],
+): Promise<ExitCode> {
   for (const blockData of executionOrder) {
     const block = blockData.block;
-    executionContext.enterNode(block);
 
-    const blockExecutor = createBlockExecutor(block);
-    const parentData = collectParents(block).map((parent) =>
-      executionOrder.find((blockData) => parent === blockData.block),
+    executionContext.enterNode(block);
+    const exitCode = await executeBlock(
+      blockData,
+      executionContext,
+      executionOrder,
     );
-    const inputValue =
-      parentData[0]?.value === undefined ? NONE : parentData[0]?.value;
-
-    let result: R.Result<IOTypeImplementation | null>;
-
-    // Check, if parent emitted a value
-    if (inputValue != null) {
-      try {
-        result = await blockExecutor.execute(inputValue, executionContext);
-      } catch (unexpectedError) {
-        executionContext.logger.logErrDiagnostic(
-          `An unknown error occurred: ${
-            unexpectedError instanceof Error
-              ? unexpectedError.message
-              : JSON.stringify(unexpectedError)
-          }`,
-          { node: blockData.block, property: 'name' },
-        );
-        return ExitCode.FAILURE;
-      }
-
-      if (R.isErr(result)) {
-        executionContext.logger.logErrDiagnostic(
-          result.left.message,
-          result.left.diagnostic,
-        );
-        return ExitCode.FAILURE;
-      }
-
-      blockData.value = result.right;
-
-      // If parent emitted no value, skip all downstream blocks
-    } else {
-      blockData.value = null;
-      executionContext.logger.logInfoDiagnostic(
-        `Skipped execution because parent block ${
-          parentData[0] ? parentData[0].block.name : 'NAME NOT FOUND'
-        } emitted no value.`,
+    executionContext.exitNode(block);
```

Comment:
I noticed that in case of failure we did not exit the node in the execution context. Was this by accident or on purpose? Let me know if I should turn it back to the old state in the latter case.